Adopt Eslint and Prettier for servicebus

KarishmaGhiya opened the pull request:
- Migrating from Tslint to ESlint
- Using Eslint for semantic checks
- Using prettier for styling checks

Some TSlint rules that are probably not supported in ESLint:
- `"typedef": [true, "call-signature", "parameter", "member-variable-declaration"]`
-  `"typedef-whitespace": [
true,
{
"call-signature": "nospace",
"index-signature": "nospace",
"parameter": "nospace",
"property-declaration": "nospace",
"variable-declaration": "nospace"
},
{
"call-signature": "onespace",
"index-signature": "onespace",
"parameter": "onespace",
"property-declaration": "onespace",
"variable-declaration": "onespace"
}
]`
- `"no-construct": true, (https://palantir.github.io/tslint/rules/no-construct/)` --> tried using https://eslint.org/docs/rules/no-useless-constructor
- "variable-name": [true, "check-format", "allow-leading-underscore", "ban-keywords"]`
- "jsdoc-format" https://palantir.github.io/tslint/rules/jsdoc-format/ => https://eslint.org/docs/rules/multiline-comment-style [For alignment of *]
Can use the plugin https://www.npmjs.com/package/eslint-plugin-jsdoc#eslint-plugin-jsdoc too but only one rule [alignment *] applicable from the available options which is similar to the rule in eslint
`"multiline-comment-style": ["error", "starred-block"]`
- `"import-blacklist": [true, "rhea", ["rhea/.*"]]` 

Following TSlint rules are probably better in eslint : 
- Promise plugin recommended (https://www.npmjs.com/package/eslint-plugin-promise)
-  Brace-style (1tbs) better than and covers all options of one-line
-  eslint-plugin-promise better than tslint no-floating-promises
- Following in tslint replaced
	`"whitespace": [
	true,
	"check-branch",
	"check-decl",
	"check-operator",
	"check-module",
	"check-seperator",
	"check-type"
	]
	}`
	Above replaced by eslint rules:	
	`"spaced-comment": ["error", "always"],
	"comma-spacing": ["error", { "before": false, "after": true }],
	"semi-spacing": ["error", {"before": false, "after": true}],
	"space-infix-ops":["error", {"int32Hint": false}],
	"keyword-spacing":["error", { "after": true }]`

Also when I run the eslint configurations set in .eslintrc.json against the current code, it generates a lot of errors and warnings : `721 problems (287 errors, 434 warnings)` . We can discuss if we want to update any settings in the .eslintrc.json.

@bterlson @ramya-rao-a @Azure/azure-sdk-eng

bterlson commented:
* Getting an error when running lint when you have d.ts files on disk. Could fix by being more specific with the paths eslint runs on (e.g. src, test, and samples).
* no-dupe-class-members should be disabled, it seems to be warning about overloads.
* eq-eq-eq should be configured to allow double-equals when one operand is null or undefined.
* I'm personally ok with explicit any as there are many cases where this is fine to use in internal implementations. Consider disabling unless someone wants this check?
* promise-always-return seems to be flagging code that is not an issue? I can't fix the errors it's giving for e.g. in `src/receiver.ts`.
* I think we could enable `allowExpressions` for explicit-function-return-type. I don't like this rule in general, but I would be happier if we didn't also have to type expressions. I think this aligns with how tslint worked?

Otherwise looks good to me!

KarishmaGhiya commented on package.json:
```diff
@@ -111,8 +111,10 @@
     "@types/node": "^10.10.0",
     "@types/nodegit": "^0.22.5",
-    "@types/yargs": "^11.0.0",
+    "@types/yargs": "^12.0.1",
```
I don't think so. I'll revert that

KarishmaGhiya commented:
> * promise-always-return seems to be flagging code that is not an issue? I can't fix the errors it's giving for e.g. in `src/receiver.ts`.

@bterlson  This rule is enabled by the recommended rules under the `promise plugin` for eslint
To fix the errors, it requires a `return` statement for every branch `if-else` as well:
Example:
`  if (!this._isClosed) {
          this._messageSession.receive(onMessage, onError, options);
        } else {
          await this._messageSession.close();
        }
      })`
The error in this snippet can be resolved by:
`   if (!this._isClosed) {
          this._messageSession.receive(onMessage, onError, options);
          return;
        } else {
          await this._messageSession.close();
          return;
        }
      })`

If this leads to unexpected behaviours, we can disable that rule.

bterlson commented:
I don't think it leads to unexpected behaviors per se, but the rule seems to flag code that actually OK. I'm also concerned that we won't be able to figure out how to make the rule happy - I tried putting returns in a bunch of places and couldn't get it to stop throwing an error. That said, happy to wait and see what others think!

KarishmaGhiya commented on sdk/servicebus/service-bus/package.json:
```diff
@@ -88,7 +94,7 @@
     "extract-api": "tsc -p . && api-extractor run --local",
     "format": "prettier --write --config .prettierrc.json \"src/**/*.ts\" \"test/**/*.ts\" \"*.{js,json}\"",
-    "lint": "tslint -p . -c tslint.json",
+    "lint": "npx eslint -c .eslintrc.json . --ext .ts",
```
Done.

bsiegel commented on sdk/servicebus/service-bus/package.json:
```diff
@@ -46,6 +46,8 @@
     "@types/node": "^8.0.37",
     "@types/ws": "^6.0.1",
+    "@typescript-eslint/eslint-plugin": "^1.7.0",
+    "@typescript-eslint/parser": "^1.7.0",
```
It looks like a lot of eslint dependencies are missing from here.

KarishmaGhiya replied:
I was having a rebase nightmare yesterday. This might be a result of that

KarishmaGhiya replied:
Fixed the missing dependencies

bsiegel commented on sdk/servicebus/service-bus/.prettierrc.json:
```diff
@@ -1,6 +1,9 @@
-  "singleQuote": false,
   "arrowParens": "always",
-  "printWidth": 100
+  "bracketSpacing": true,
```
If we think this is a good prettier config for all of our projects, should we store it in the root of the repo and symlink it into each package's directory?

KarishmaGhiya replied:
done

bsiegel commented on sdk/servicebus/service-bus/.eslintrc.json:
```diff
@@ -0,0 +1,86 @@
+{
```
I'd like to poke at this and see if I can get the list of rules reduced - ideally we should be as close to `eslint:recommended` as possible, right?

Also, if we think this is a good eslint config for all of our projects, is it possible to store this in the root of the repo and then reference it in each package like the following example:

.eslintrc.json:
```
{
    "extends": ["azure-sdk-for-js"]
}
```

I wouldn't do the symlinking thing here, since each package may need to actually override some of our common rules (unlike Prettier which is designed to be consistent across everything that uses it).

KarishmaGhiya replied:
@bsiegel  The eslint rules in the PR are basically conversion of the existing tslint rules configured for the package. I did cross-verify to see if i have any unnecessary conflicting rules with those covered by prettier and removed those. But whether we would want to revise for the other rules and about standardizing across all packages, I will discuss this with @bterlson

bsiegel replied:
Sounds good! I'm not 100% sure of the mechanism, but I'm reasonably sure we can keep a central copy of our "best practice" lint rules, and then `extend` from them in each library's .eslintrc.json, overriding or adding specific rules as needed for the library.

Regardless, that should definitely be its own task, I wouldn't argue you should try to do that in this PR.

bsiegel commented on sdk/servicebus/service-bus/package.json:
```diff
@@ -44,21 +44,29 @@
     "@types/dotenv": "^6.1.0",
     "@types/mocha": "^5.2.5",
-    "@types/node": "^8.0.0",
+    "@types/node": "^8.0.37",
```
Please revert this change, this should be pinned at ^8.0.0

KarishmaGhiya replied:
done

jeremymeng commented on sdk/servicebus/service-bus/src/util/utils.ts:
```diff
@@ -143,7 +143,7 @@ export function toBuffer(input: any): Buffer {
     // coercing undefined to null as that will ensure that null value will be given to the
     // customer on receive.
-    if (input === undefined) input = null; // tslint:disable-line
+    if (input === undefined) input = null; // eslint:disable-curly
```
I'd rather have this fixed than disabling the rule

KarishmaGhiya replied:
fixed

bterlson commented:
Took a look through again, looking good. Some more thoughts:

* Need to allow `==` with `null` and `undefined`.
* IMO Curly should be configured with `multi-line` so `if (test) doThing()` is an allowed pattern.
* Does the pkg json ordering linter do what Rush does (just want to make sure there aren't conflicts and we shouldn't get lint errors after running rush update on a clean repo).

bsiegel commented:
> * IMO Curly should be configured with `multi-line` so `if (test) doThing()` is an allowed pattern.

It looks like there are some conflicts between eslint and prettier when `multi-line` is used: https://github.com/prettier/eslint-config-prettier#curly

The example they provide in the docs there is actually a great motivating example of situations in general that make me suspicious of allowing curlies to be omitted.

bterlson commented:
I would say in general, if you have a line break after your `if`, you should have a block (and prettier's formatting rules are such that you might end up with a linebreak if you have a complex test), but otherwise it's not necessary. To give an explicit example, I believe this:

```
if (cancelled) return;
```

is more clear or is not more error prone than

```
if (cancelled) {
  return;
}
```

I'm ok if prettier's formatting will produce code linter will complain about later (prettier won't remove the braces once you add them in IIUC?). But I won't go to the mats for this one.

weshaggard commented on sdk/servicebus/service-bus/.prettierrc.json:
```diff
@@ -1,6 +1,9 @@
 {
```
Once we get this the rules worked out I think this should live at the scope of where all the source code that will share it lives, so I think that would put it at `sdk/.prettierrc.json`.

weshaggard replied:
I think the eslintrc.json file should likely be similarly scoped.

KarishmaGhiya replied:
Thank you @weshaggard

KarishmaGhiya commented:
> Need to allow == with null and undefined.
@bterlson 
I can't see to find an option with `eqeqeq` that allows `undefined` .The following allows `null` but not `undefined` with `eqeqeq`. The option `allow-null` that used to support both is deprecated. 
There's an option `smart` with the rule eqeqeq that allows code like the following:
`typeof foo == 'undefined'` Do you think we would want to use that? What are your thoughts?
Refer : https://eslint.org/docs/rules/eqeqeq

KarishmaGhiya commented:
> Does the pkg json ordering linter do what Rush does (just want to make sure there aren't conflicts and we shouldn't get lint errors after running rush update on a clean repo).

@bterlson I removed the options that I had added to run linting on package.json for sorting the keys because it was not serving the purpose we want. Using the eslint rules the only way to enable the sorting for only specific parts of the file is by using comments. But comments are not allowed in json files. The other options I came across are -
- Using some external package that are specific for package.json like fixpack
https://www.npmjs.com/package/fixpack
- Using the `sort lines` options under command pallette in vscode or sublime text and let the developer sort it manually

KarishmaGhiya commented:
/azp run azure-sdk-for-js - client

bsiegel commented:
Let's also codify some of these rules in the .vscode/settings.json file:

```diff
diff --git a/.vscode/settings.json b/.vscode/settings.json
index be74596e..3793787b 100644
--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -1,3 +1,8 @@
 {
+  "editor.autoIndent": true,
+  "editor.formatOnSave": true,
+  "editor.tabSize": 2,
+  "files.insertFinalNewline": true,
+  "files.trimFinalNewlines": true,
   "typescript.disableAutomaticTypeAcquisition": true
-}
\ No newline at end of file
+}
```

bterlson commented:
I guess we really want eqeqeq so have to take some pain. Let's roll with allowing null ("always", null: ignore) and can update all `== undefined` to `== null`.

bsiegel commented on sdk/.eslintrc.json:
```diff
@@ -0,0 +1,77 @@
+  "rules": {
+    "curly": ["error", "multi-line"],
+    "eol-last": ["error", "always"],
+    "eqeqeq": ["error", "always",{"null": "ignore"}],
```
nit: space before bracket
